src/System.CommandLine/Parsing/CommandResult.cs

```diff
@@ -36,32 +36,147 @@ public sealed class CommandResult : SymbolResult
         /// </summary>
         public IEnumerable<SymbolResult> Children => SymbolResultTree.GetChildren(this);
 
-        internal sealed override int MaximumArgumentCapacity
+        internal override bool UseDefaultValueFor(ArgumentResult argumentResult)
+            => argumentResult.Argument.HasDefaultValue && argumentResult.Tokens.Count == 0;
+
+        /// <param name="completeValidation">Only the inner most command goes through complete validation.</param>
+        internal void Validate(bool completeValidation)
```

Comment:
I've moved this logic from the visitor and re-wrote in a way that we don't iterate over every argument more than once (previously we were setting the default values and then validating).